### PR TITLE
clock_control: L4/L0: Enable MSI as possible direct SYSCLK source on STM32 L0/L4

### PR DESCRIPTION
**Pull request by erwango**
MSI could be useful to provide an accurate clock when no external oscillator is available.

Add this option in STM32 clock_control driver.
Modify disco_l475_iot1 Kconfig.defconfig in order to propose 2 alternate default configurations.
MSI driven SYSCLK configuration is useful in BLE cases where accurate clock is needed.

### CODE REVIEW

File: boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig [hunk 1]
```diff
@@ -17,16 +16,15 @@ CONFIG_GPIO=y
...
 CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
-# produce 80MHz clock at PLL output
-CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
-CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=20
-CONFIG_CLOCK_STM32_PLL_P_DIVISOR=7
-CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
-CONFIG_CLOCK_STM32_PLL_R_DIVISOR=4
+# Alternative for BLE use cases: Direct HSI
+# CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
+
```
Comment (ydamigos) [inline boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig]: MSI?

File: boards/arm/disco_l475_iot1/Kconfig.defconfig [hunk 2]
```diff
@@ -10,6 +10,33 @@ if BOARD_DISCO_L475_IOT1
...
+	default 1
+config CLOCK_STM32_PLL_N_MULTIPLIER
+	default 20
+config CLOCK_STM32_PLL_P_DIVISOR
+	default 7
+config CLOCK_STM32_PLL_Q_DIVISOR
+	default 2
+config CLOCK_STM32_PLL_R_DIVISOR
+	default 4
+endif # SYS_CLOCK_HW_CYCLES_PER_SEC=80000000
```
Comment (ydamigos) [inline boards/arm/disco_l475_iot1/Kconfig.defconfig]: Please remove.

File: boards/arm/disco_l475_iot1/Kconfig.defconfig [hunk 3]
```diff
@@ -10,6 +10,33 @@ if BOARD_DISCO_L475_IOT1
...
+
+endif # CLOCK_STM32_PLL_SRC_HSI
```
Comment (ydamigos) [inline boards/arm/disco_l475_iot1/Kconfig.defconfig]: Move this after
```
config SYS_CLOCK_HW_CYCLES_PER_SEC
	default 16000000
```

**Review by ydamigos — CHANGES_REQUESTED** [review]
There is a mix with the `if` / `endif` in Kconfig.defconfig

**Comment by erwango** [comment]
@ydamigos : thanks for review, I pushed too quickly. Should be better now

File: boards/arm/disco_l475_iot1/Kconfig.defconfig [hunk 4]
```diff
@@ -57,4 +86,6 @@ config VL53L0X_XSHUT_GPIO_PIN_NUM
 
 endif #VL53L0X
 
+
+
```
Comment (ydamigos) [inline boards/arm/disco_l475_iot1/Kconfig.defconfig]: Could you remove this lines?

File: boards/arm/disco_l475_iot1/Kconfig.defconfig [hunk 5]
```diff
@@ -10,6 +10,35 @@ if BOARD_DISCO_L475_IOT1
...
+endif # CLOCK_STM32_PLL_SRC_HSI
+
+endif # CLOCK_STM32_SYSCLK_SRC_PLL
+
+if CLOCK_STM32_SYSCLK_SRC_MSI
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 16000000
+
+endif # CLOCK_STM32_SYSCLK_SRC_PLL
```
Comment (ydamigos) [inline boards/arm/disco_l475_iot1/Kconfig.defconfig]: CLOCK_STM32_SYSCLK_SRC_MSI?

**Review by ydamigos — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
@ydamigos : Done, thanks!

**Review by ydamigos — APPROVED** [review]
@erwango Thanks for the changes.

**Review by galak — APPROVED** [review]

**Pull request merged**